tests/std/tests/P3349R1_contiguous_iterators_to_pointers/test.cpp

```diff
@@ -0,0 +1,506 @@
+
+#undef TESTW
+#undef TESTR
+#undef EXPR_WRAP_IS_SMART
+    }
+
+    // i + [0, n)
+    {
+#define TESTR(expr) tests_base(EXPR_WRAP_IN(expr), r_first_m, r_dist_m)
+#define TESTW(expr) tests_base(EXPR_WRAP_IN(expr), w_first_s, w_dist_s)
+
+        TESTR(copy_n(i, n, w_first_l));
+        TESTR(ranges::copy_n(i, n, w_first_l));
+
+        TESTW(fill_n(i, n, container_read[0]));
+        TESTW(ranges::fill_n(i, n, container_read[0]));
+
+        TESTR(uninitialized_copy_n(i, n, w_first_l));
+        TESTR(ranges::uninitialized_copy_n(i, n, w_first_l, unreachable_sentinel));
+        tests_base(EXPR_WRAP_IS(ranges::uninitialized_copy_n(r_first_m, r_dist_m, i, s)), w_first_s, w_last_s);
+
+        TESTW(uninitialized_move_n(i, n, w2_first_l));
+        TESTW(ranges::uninitialized_move_n(i, n, w2_first_l, unreachable_sentinel));
+        tests_base(EXPR_WRAP_IS(ranges::uninitialized_move_n(w_first_s, w_dist_s, i, s)), w2_first_s, w2_last_s);
+
+        TESTW(uninitialized_fill_n(i, n, container_read[0]));
+        TESTW(ranges::uninitialized_fill_n(i, n, container_read[0]));
+
+#undef TESTW
+#undef TESTR
+    }
+}
+
+template <bool Nothrow>
+void test_matrix() {
+    // Ensures vectorization (and iterator-to-pointer conversion)
+    test<int8_t, 63, Nothrow>();
+    test<int16_t, 256, Nothrow>();
+    test<int32_t, 256, Nothrow>();
+    test<int64_t, 256, Nothrow>();
+
+    // remove_copy and unique_copy
+    {
+        using iter       = safe_iter<int, Nothrow>;
+        using const_iter = safe_iter<const int, Nothrow>;
+
+        const auto src = {1, 2, 3, 4, 4, 5, 5};
+
+        const auto src_first    = const_iter::begin(src);
+        const auto src_last     = const_iter::end(src);
+        const auto src_last_ptr = to_address(src_last);
+
+        array<int, 5 + 1> dst;
+        const iter dst_first{
+            to_address(ranges::begin(dst)), to_address(ranges::begin(dst)), to_address(ranges::begin(dst) + 5)};
+
+        PASS(remove_copy(src_first, src_last, dst_first, 5));
+        FAIL(remove_copy(src_first, src_last, dst_first, 1));
+        PASS(ranges::remove_copy(src_first, src_last_ptr, dst_first, 5));
+        FAIL(ranges::remove_copy(src_first, src_last_ptr, dst_first, 1));
+
+        PASS(unique_copy(src_first, src_last, dst_first));
+        FAIL(unique_copy(src_first, src_last, dst_first + 1));
+        PASS(ranges::unique_copy(src_first, src_last_ptr, dst_first));
+        FAIL(ranges::unique_copy(src_first, src_last_ptr, dst_first + 1));
+    }
+
+    // search_n
+    {
+        using const_iter = safe_iter<const int, Nothrow>;
+
+        const auto rng             = {1, 1, 2, 3, 3, 3};
+        const auto valid_first_ptr = rng.begin() + 1;
+        const auto valid_last_ptr  = rng.end() - 1;
+
+        const const_iter rng_first{valid_first_ptr, valid_first_ptr, valid_last_ptr};
+        const const_iter rng_last{valid_last_ptr, valid_first_ptr, valid_last_ptr};
+        const auto rng_last_ptr = to_address(rng_last);

```

Comment:
I don't understand the purpose of `rng_last_ptr` - it's just `valid_last_ptr` and this isn't exercising any product code.